### PR TITLE
Multiple splits

### PR DESCRIPTION
**Pull request by asbjrnmunk**
Currently, yucca only supports one type of split per task, split manager combination. However i would like to make some experiments where i vary the amount of data in my train split. This changes the `split_config` such that it stores splits as a dict of dicts of lists of dicts instead of before where it was a list of dicts:

before:
```
[{'train': [...], 'val': [...]}, ....]
```

after:
```
{ 'method': { 'param': [{'train': [...], 'val': [...]}, ....] } }
```
where param is either the `k` or `p` value, uniqly identified by the method.

Example:

```
{ 
    'kfold': { '5': [{'train': [...], 'val': [...]}, ....] },
    'simple_train_val_spit': { '0.1': [{'train': [...], 'val': [...]}, ....] }
}
```

Importantly, we only compute the split when it is needed and store it so we only have to compute it once.

### CODE REVIEW

**Comment by jakobamb** [comment]
tests are still failing, but generally looks good!

The option to supply either k or p split is inherently awkward, so I think we should catch cases where both k and p are supplied directly in run_training.py as well, before starting up the whole machine until finally erroring in the actual function.

**Review by Sllambias — COMMENTED** [review]
I like the solution, I do however think we could simplify it even more to have only one of K or P. Probably by scrapping the K argument. So we have a Method and a Parameter for that method, so for KFOLD with Param = 5 we have 5-Fold. 
Then we avoid the either/or situations just have a straightforward API I think

**Comment by Sllambias** [comment]
My suggestion would be:
```
 parser.add_argument(
      "--split_param",
      type=Union[float, int],
      help="Can be either Integer or Float. If split_param is an integer > 1 K-Fold splits are created with `split_param` == K. If split_param is a float 1>split_param>0 we use a train/val split where `split_param` is the fraction of items used for the val split.",
      default=5,
  )
```

**Comment by asbjrnmunk** [comment]
@jakobamb what do you think about Sebastians suggestion? 

My gut reaction is no, however I must say that I agree that magic `k` and `p` params are also not optimal. This is also magical, and maybe even more magical because we branch on types and assert different parameter ranges for each type, however i do like that it has the word "split" in the api.

**Comment by asbjrnmunk** [comment]
Yes, I think this is the right way to go. @Sllambias what do you think?

**Comment by Sllambias** [comment]
Yeah, not a huge fan of either aswell. I think this is fine by me as long as it still works to not supply anything and it defaults to the same as usual. Now we just need to handle paths: How are models saved when they're using 
1. 5fold and trained on fold 0
2. 5fold and trained on fold 2
3. simple_split trained on fold/split 0
4. simple_splt trained on fold/split 2 (is this actually possible, or will it always be 0?)

right now its 
`    save_dir = join(
        yucca_models,
        task,
        model_name + "__" + model_dimensions,
        manager_name + "__" + planner,
        experiment,
        f"fold_{split_idx}",
    )
`

but I'm afraid 1 and 3 above will yield the same save path

**Comment by Sllambias** [comment]
So the slight challenge I realize here is that we set path before doing the splits config, so we can't access the params of the SplitConfig for this, so we must rely on the parsed args. 

Maybe a solution could be:

`
if K is not None:
    path = f"{K}Fold_{split_idx}"
if P is not None:
    path = f"SimpleSplit_{split_idx}"
`

what do u think?

**Comment by asbjrnmunk** [comment]
I al looking at it :-)

**Comment by asbjrnmunk** [comment]
@Sllambias i think this is the most clean solution: we treat `split_method` and `split_param` in the same way as `split_idx` and include it in the `task_config`.

**Comment by asbjrnmunk** [comment]
Do you know if checkpointing still works when the `split_config.lm_hparams` contain `split_method` and `split_param` instead of `spliit_data_ratio` and `split_data_kfold`?

File: yucca/run/run_training.py [hunk 1]
```diff
@@ -113,6 +118,17 @@ def main():
     profile = args.profile
     experiment = args.experiment
 
+    split_idx = args.split_idx
+    split_data_ratio = args.split_data_ratio
+    split_data_kfold = args.split_data_kfold
+
+    if split_data_kfold is None and split_data_ratio is None:
+        split_data_kfold = 5
```
Comment (Sllambias) [inline yucca/run/run_training.py:126]: Does this also need to be added to run_finetune?

File: yucca/training/configuration/configure_paths.py [hunk 2]
```diff
@@ -27,20 +28,26 @@ def lm_hparams(self):
 
 
 def get_path_config(task_config: TaskConfig):
-    save_dir, train_data_dir, version_dir, plans_path, version = setup_paths_and_version(
-        task_config.continue_from_most_recent,
-        task_config.manager_name,
-        task_config.model_dimensions,
-        task_config.model_name,
-        task_config.split_idx,
+    task_dir = join(yucca_preprocessed_data, task_config.task)
```
Comment (Sllambias) [inline yucca/training/configuration/configure_paths.py:31]: do we use this task_dir for anything that isnt covered by train_data_dir?

Comment (Sllambias) [reply yucca/training/configuration/configure_paths.py:31]: nvm, found it myself ^^

**Review by Sllambias — APPROVED** [review]
lovely, lgtm

File: yucca/run/run_training.py (hunk 1 above)
Comment (asbjrnmunk) [reply yucca/run/run_training.py:126]: yes

**Pull request merged**